ntoskrnl/se/token.c

```diff
@@ -135,34 +135,232 @@ SepDeleteTokenLock(
+    for (FirstCount = 0; FirstCount < CountPrivArray1; FirstCount++)
+    {
+        for (SecondCount = 0; SecondCount < CountPrivArray2; SecondCount++)
+        {
+            if (PrivArrayToken1[FirstCount].Luid.LowPart !=
+                PrivArrayToken2[SecondCount].Luid.LowPart)
+            {
+                DPRINT("SepComparePrivilegeAndAttributesFromTokens(): Privileges are not equal!\n");
+                continue;
+            }
+
+            if (PrivArrayToken1[FirstCount].Attributes !=
+                PrivArrayToken2[SecondCount].Attributes)
+            {
+                DPRINT("SepComparePrivilegeAndAttributesFromTokens(): Privilege attributes are not equal!\n");
+                continue;
+            }
+
+            break;
+        }
+
+        /* We've exhausted the array of the second token without finding this one */
+        if (SecondCount == CountPrivArray2)
+        {
+            DPRINT("SepComparePrivilegeAndAttributesFromTokens(): No matching elements could be found in either token!\n");
+            return FALSE;
+        }
+    }
+
+    return TRUE;
+}
+
+/**
+ * @brief
+ * Compares tokens if they're equal based on all the following properties. If all
+ * of the said conditions are met then the tokens are deemed as equal.
+ *
+ * - Every SID that is present in either token is also present in the other one.
+ * - Both or none of the tokens are restricted.
+ * - If both tokens are restricted, every SID that is restricted in either token is
+ *   also restricted in the other one.
+ * - Every privilege present in either token is also present in the other one.
+ * 
+ * @param[in] FirstToken
+ * The first token.
+ *
+ * @param[in] SecondToken
+ * The second token.
+ *
+ * @param[out] Equal
+ * The retrieved value which determines if the tokens are
+ * equal or not.
+ *
+ * @return
+ * Returns STATUS_SUCCESS.
+ */
+static
+NTSTATUS
+SepCompareTokens(
+    _In_ PTOKEN FirstToken,
+    _In_ PTOKEN SecondToken,
+    _Out_ PBOOLEAN Equal)
 {
     BOOLEAN Restricted, IsEqual = FALSE;
+    PAGED_CODE();
 
     ASSERT(FirstToken != SecondToken);
 
     /* Lock the tokens */
     SepAcquireTokenLockShared(FirstToken);
     SepAcquireTokenLockShared(SecondToken);
 
-    /* FIXME: Check if every SID that is present in either token is also present in the other one */
+    /* Check if every SID that is present in either token is also present in the other one */
+    if (!SepCompareSidAndAttributesFromTokens(FirstToken->UserAndGroups,
+                                              FirstToken->UserAndGroupCount,
+                                              SecondToken->UserAndGroups,
+                                              SecondToken->UserAndGroupCount))
+    {
+        goto Quit;
+    }
 
+    /* Are both tokens restricted? */
     Restricted = SeTokenIsRestricted(FirstToken);
     if (Restricted == SeTokenIsRestricted(SecondToken))
     {
         if (Restricted)
         {
```

Comment:
As a side comment, it would appear that these two if-s could be merged into a single one:
```
```suggestion
    Restricted = SeTokenIsRestricted(FirstToken);
    if (Restricted && (Restricted == SeTokenIsRestricted(SecondToken)))
    {
```